resources/js/features/user-game-activity/components/UserGameActivityTimeline/SessionTimelineEvent/UnlockEventContent.tsx

```diff
@@ -0,0 +1,106 @@
+import { getIsAchievementPublished } from '@/common/utils/getIsAchievementPublished';
+import { useFormatDuration } from '@/common/utils/l10n/useFormatDuration';
+
+interface UnlockEventContentProps {
+  achievement: App.Platform.Data.Achievement;
+  hardcore: boolean;
+  hardcoreLater: boolean;
+  previousEventKind: App.Enums.PlayerGameActivityEventType | 'start-session';
+  when: string | null;
+  whenPrevious: string | null;
+}
+
+export const UnlockEventContent: FC<UnlockEventContentProps> = ({
+  achievement,
+  hardcore,
+  hardcoreLater,
+  previousEventKind,
+  when,
+  whenPrevious,
+}) => {
+  const isOfficialAchievement = getIsAchievementPublished(achievement);
+
+  return (
+    <AchievementAvatar
+      {...achievement}
+      imgClassName={cn(isOfficialAchievement ? null : 'grayscale')}
+      showHardcoreUnlockBorder={hardcore}
+      showPointsInTitle={true}
+      sublabelSlot={
+        <AchievementTimingLabel
+          achievement={achievement}
+          hardcore={hardcore}
+          hardcoreLater={hardcoreLater}
+          previousEventKind={previousEventKind}
+          when={when}
+          whenPrevious={whenPrevious}
+        />
+      }
+    />
+  );
+};
+
+interface AchievementTimingLabelProps {
+  achievement: App.Platform.Data.Achievement;
+  hardcore: boolean;
+  hardcoreLater: boolean;
+  previousEventKind: App.Enums.PlayerGameActivityEventType | 'start-session';
+  when: string | null;
+  whenPrevious: string | null;
+}
+
+const AchievementTimingLabel: FC<AchievementTimingLabelProps> = ({
+  achievement,
+  hardcore,
+  hardcoreLater,
+  previousEventKind,
+  when,
+  whenPrevious,
+}) => {
+  const { t } = useTranslation();
+
+  const { formatDuration } = useFormatDuration();
+
+  const isOfficialAchievement = getIsAchievementPublished(achievement);
+
+  const diffInSeconds = when && whenPrevious ? dayjs(when).diff(dayjs(whenPrevious), 'second') : 0;
+
+  // No label is needed for instant achievements at session start.
+  if (!hardcoreLater && previousEventKind === 'start-session' && diffInSeconds === 0) {
+    return null;
+  }
+
+  return (
+    <span className="flex items-center gap-0.5 text-2xs text-neutral-500">
+      <LuArrowRight />
+
+      {!hardcore ? <span className="mr-1 text-neutral-500">{t('(Softcore)')}</span> : null}
+
+      {!isOfficialAchievement ? (
+        <span className="mr-1 font-semibold text-neutral-500">{t('Unofficial Achievement')}</span>
+      ) : null}
+
+      {hardcoreLater ? t('Unlocked later in hardcore') : null}
+
+      {previousEventKind === 'start-session' && diffInSeconds > 0
+        ? t('{{time}} after session start', {
+            time: formatDuration(diffInSeconds),
+          })
+        : null}
+
+      {previousEventKind !== 'start-session' &&
+        (diffInSeconds > 0
+          ? t('{{time}} after previous', {
+              time: formatDuration(diffInSeconds),
+            })
+          : t('Immediately after previous'))}
```

Comment:
For something like this:
![image](https://github.com/user-attachments/assets/58b449c4-9a7d-40ed-876d-a0bff8bb7733)

Would it make sense to group them in a single cell?
![image](https://github.com/user-attachments/assets/2add4e6c-0dc7-4d79-acc4-8008f7d33a3f)

I understand it might be more difficult than it's worth. Maybe just leave off (or gray out) the redundant timestamp?

Reply:
I think it's a good idea. Done in latest:

![Screenshot 2025-01-01 at 2 57 22 PM](https://github.com/user-attachments/assets/4a21f2f3-71f4-47dc-b228-bd7c5e360642)